### PR TITLE
Add iOS simulator fallback for UTType

### LINKED ISSUE #16383: [expo-media-library] createAssetAsync() on iOS throws: This file type is not supported yet
### Summary

I'm trying to add an image captured with `react-native-view-shot` into photos. Android performs as expected, but iOS throws `This file type is not supported yet` error. The required file type extension seems to be in place, the `uri` is something like this:

`/Users/***/Library/Developer/CoreSimulator/Devices/1C98873B-B902-40D7-BFBE-47810F8662E3/data/Containers/Data/Application/15D1F710-C902-40DC-B910-731BA20971F2/tmp/ABI44_0_0ReactNative/30EF0F17-BB12-4D8B-AFD5-DB74845C9696.jpg`

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

iOS

### SDK Version (managed workflow only)

44

### Environment

  expo-env-info 1.0.2 environment info:
    System:
      OS: macOS 12.1
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
      Yarn: 1.22.17 - /opt/homebrew/bin/yarn
      npm: 8.1.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
      Watchman: 2021.11.15.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.2 - /opt/homebrew/lib/ruby/gems/3.0.0/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.0.1, iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0
      Android SDK:
        API Levels: 29, 30
        Build Tools: 29.0.2, 30.0.2
        System Images: android-29 | Google APIs ARM 64 v8a, android-29 | Google APIs Intel x86 Atom_64, android-31 | ARM 64 v8a, android-31 | Google Play ARM 64 v8a
    IDEs:
      Android Studio: 2021.1 AI-211.7628.21.2111.8005941
      Xcode: 13.1/13A1030d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~44.0.0 => 44.0.6 
      react: 17.0.1 => 17.0.1 
      react-dom: 17.0.1 => 17.0.1 
      react-native: 0.64.3 => 0.64.3 
      react-native-web: 0.17.1 => 0.17.1 
    npmGlobalPackages:
      eas-cli: 0.47.0
      expo-cli: 5.2.0
    Expo Workflow: managed

### Reproducible demo

```
// App.tsx

import { createAssetAsync, usePermissions } from "expo-media-library";
import { useRef } from "react";
import { Pressable, StyleSheet, Text, View } from "react-native";
import { captureRef } from "react-native-view-shot";

export default function App() {
  const viewRef = useRef();

  const [_, requestPermission] = usePermissions();

  const captureImage = async (ref) => {
    try {
      const permission = await requestPermission();

      if (permission.granted) {
        const uri = await captureRef(ref.current, { format: "jpg" });

        const asset = await createAssetAsync(uri);

        if (asset) {
          console.log("Success:", asset);
        }
      }
    } catch (error) {
      console.log("Fail:", error);
    }
  };

  return (
    <View ref={viewRef} style={styles.container}>
      <Pressable
        style={{ width: 100, height: 100, backgroundColor: "green" }}
        onPress={() => captureImage(viewRef)}
      />
      <Text>Open up App.tsx to start working on your app!</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
});
```

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

Closes https://github.com/expo/expo/issues/16383

It seems this line doesn't work on iOS simulator on M1 macs when the app is run through Rosetta (as Expo GO is):
```
expo/packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:927

CFStringRef fileUTI = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, fileExtension, NULL);**
```

It seems that this is a common issue:

https://github.com/ionic-team/capacitor/issues/4903

https://developer.apple.com/forums/thread/685968

https://developer.apple.com/forums/thread/702933

The `UTTypeCreatePreferredIdentifierForTag` method is also deprecated.

Unfortunately, changing to the non-deprecated `[UTType typeWithFilenameExtension]`  also returns nil on M1 simulator.

# How

We can do a lookup table for mapping extensions to media types if iOS functions return nil.

# Test Plan

I've tested the fallback code on iOS simulator.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-media-library/CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 ### 🎉 New features
 
 ### 🐛 Bug fixes
+- Fix promise rejection on M1 iOS Simulator due to `UTTypeCreatePreferredIdentifierForTag` not working as expected. ([#19669](https://github.com/expo/expo/pull/19669) by [@aleqsio](https://github.com/aleqsio))
```
Comment (lukmccall) [inline packages/expo-media-library/CHANGELOG.md:12]:
```suggestion

- Fix promise rejection on M1 iOS Simulator due to `UTTypeCreatePreferredIdentifierForTag` not working as expected. ([#19669](https://github.com/expo/expo/pull/19669) by [@aleqsio](https://github.com/aleqsio))
```
🧹

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 2]
```diff
@@ -924,7 +924,57 @@ + (NSString *)_assetUriForLocalId:(nonnull NSString *)localId
...
+    fileUTI = (__bridge CFStringRef)[[UTType typeWithFilenameExtension:(__bridge NSString*)fileExtension] identifier];
+  } else {
+    fileUTI = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, fileExtension, NULL);
+  }
+
+  // A fix for promise rejection on M1 simulator when run in Rosetta (as is by default as we don't compile ARM64 yet)
+  // https://developer.apple.com/forums/thread/702933
+  if(fileUTI == nil){
+    // List from https://stackoverflow.com/a/70102692
+    NSDictionary* kExtensionLookupFallback = @{@"jpeg": @(PHAssetMediaTypeImage),
```
Comment (lukmccall) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:938]:
```suggestion
    NSDictionary *kExtensionLookupFallback = @{@"jpeg": @(PHAssetMediaTypeImage),
```

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 3]
```diff
@@ -924,7 +924,57 @@ + (NSString *)_assetUriForLocalId:(nonnull NSString *)localId
...
+                                               @"kar": @(PHAssetMediaTypeAudio),
+                                               @"tiff": @(PHAssetMediaTypeImage),
+                                               @"tif": @(PHAssetMediaTypeImage),
+                                               @"gif": @(PHAssetMediaTypeImage),
+                                               @"qtif": @(PHAssetMediaTypeImage),
+                                               @"qti": @(PHAssetMediaTypeImage),
+                                               @"icns": @(PHAssetMediaTypeImage)};
+
+    EXLogWarn(@"Asset media type is recognized from file extension and this behavior can differ on iOS Simulator and a physical device.");
+    NSNumber* fallbackMediaType = [kExtensionLookupFallback objectForKey:[localUri pathExtension]];
```
Comment (lukmccall) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:975]:
```suggestion
    NSNumber *fallbackMediaType = [kExtensionLookupFallback objectForKey:[localUri pathExtension]];
```

**Review by lukmccall — APPROVED** [review]
Nice 🔥

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 4]
```diff
@@ -924,7 +924,57 @@ + (NSString *)_assetUriForLocalId:(nonnull NSString *)localId
...
+  CFStringRef fileUTI;
+  if (@available(iOS 14, *)) {
+    fileUTI = (__bridge CFStringRef)[[UTType typeWithFilenameExtension:(__bridge NSString*)fileExtension] identifier];
+  } else {
+    fileUTI = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, fileExtension, NULL);
+  }
+
+  // A fix for promise rejection on M1 simulator when run in Rosetta (as is by default as we don't compile ARM64 yet)
+  // https://developer.apple.com/forums/thread/702933
+  if(fileUTI == nil){
```
Comment (tsapeta) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:936]:
```suggestion
  if (fileUTI == nil){
```

Also, could you move this logic to a separate method? Something like
`+ (PHAssetMediaType)assetTypeForFileExtension:(nonnull NSString *)fileExtension` should be fine (you can omit `_` for private methods, it's a relic from the past in our codebase)

**Review by tsapeta — APPROVED** [review]

**Comment by brentvatne** [comment]
@tsapeta - can ensure that this lands before ios versioning?

**Review by brentvatne — APPROVED** [review]

**Comment by aleqsio** [comment]
Ci for check packages fails due to https://github.com/expo/expo/pull/19621 on main, so unrelated issue. Tested the new method on device.

**Pull request merged**